Performance-Tuning enhancement proposal.

alanconway opened the pull request:
Covers delivery reliability (log loss) and sundry output tuning (reconnect, compression, batching)
I'm happy that this proposal is usable and outlines a feasible implementation.

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+
+Persistent buffering is done to compensate for a `source` but happens on the `output`:
+- acknowledgement will still be used for sources that support it.
+- adding a buffer for the weak sources doesn not interfere with acknowledgement.
```
we need to word smith this.  While conceptionally it is accurate, buffering is added to outputs.  We need to be clear that we are not adding "source" buffers because we have no possibility to make that a reality

alanconway replied:
Agreed,  part of rework for your other comments above.

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+- Support cost of customers abusing or misunderstanding the parameters. \
+  For `delivery` this is mitigated by having the forward compute a "correct" config.
+- Increased customer demand for help "sizing" the logging stack: setting resources and predicting performance.
+- Now new security risks are expected.
```
What security risks does this change add?

jcantrill reviewed (commented):
While I don't wish to add a field for every vector tuning option, I don't think we can capture all tuning in the way described here.  I believe we are really going need to allow a user to explicitly define a parameter like blocking behavior or max events.  I don't see any option other then explicit tuning and we have customers who are already experiencing this short fall

alanconway commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+
+#### delivery: Fast|Safe
+
+`Fast`: Minimize resource use (CPU, disk, network) and maximize throughput in normal operation. Use when:
```
I mean minimize resources at a given throughput. I.e. if the sustained log load is X Mb/sec, and that is low enough not to overload either fast or safe mode, then fast mode should use less resources. If you push either mode to it's limits then fast mode can use as much resource as safe mode, but the throughput limit for fast mode should be higher before it gets  overloaded. I'll think about re-wording here, it is an important point to distinguish between "normal operation" and "pushing the system to overload".

alanconway replied:
Reworded, please check

alanconway commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+- All outputs accept `compress: true` to enable a default mode.
+- Error if output does not support compression or does not recognize the value.
+
+#### batch: object
```
Fair point, I'll review. I think fluent has equivalent "chunk" settings but I'll make sure. Also we can restate these values as being "desired" or "soft" limits that may not be respected exactly depending on the underlying technology. E.g. not all protocols support any kind of compression at etc. Probably we should raise an error if if it is not possible to respect any perf parameter so there are no false expectations.

alanconway replied:
I'll drop these from the initial proposal, we can bring them back when we have  a use case.

jcantrill replied:
> I missed the fact there is a batch section for vector so my comment may no longer be accurate
From my slack comment.  Re-evaluating....

alanconway commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+- `maxRecords`: Max number of records to include in a batch.
+- `maxBytes`: Max size in bytes of a batch
+
+#### reconnect: object
```
+1, I'll review.

alanconway replied:
Dropped initially, we can revisit later.

alanconway commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+#### Acknowledgements
+
+[End-to-end acknowledgement](https://vector.dev/docs/about/under-the-hood/architecture/end-to-end-acknowledgements) 
+means that source acknowledgements are _delayed_ until all relevant sinks have received the data.
```
I will look at that more carefully. The proper combo of acks, buffers and other features (rotate_wait, drop_oldest) may vary by source and sink. I'll put together some notes for implementation.

alanconway replied:
The file-ack is still needed for collector restarts, rotate_wait doesn't affect that (nor does file-ack affect rotate_wait)

alanconway commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+This is more efficient than duplicating the data again in a vector disk buffer,
+but only works for sources that support acknowledgement and "at-least-once" reliable delivery.
+
+Examples of acknowledgement sources:
```
End-to-end acknowledgement involves both the source and the sink. The source defines how vector acknowledges to the sink that it can "forget" data that has been sucessfully sent. E.g. for HTTP it's a HTTP response, for a file source it is moving the persistent "position". The sink is involved because it receives ack that the data was sucessfully sent. The source delays ack to the remote sink until all relevant sources have indicated that they have sent OK. All sinks support end-to-end ack, even if the sink protocol is not "reliable" - they report when the data is sent "to the best of their knowledge". Only some sources support it,  it requires that the remote source will "re-send" unacked messages. In the case of files this just means vector will re-read the file, for network protcols it's an actual re-send after re-connect. I'll clarify in the doc.

alanconway commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,238 @@
+- All outputs accept `compress: true` to enable a default mode.
+- Error if output does not support compression or does not recognize the value.
+
+#### batch: object
```
Logically separate concern. Persistent buffer is for recovery from disconnects. Batch size is max data per individual "write". Will review vector buffering to see if they clash in the implementation...

jcantrill replied:
My point to mention it here is that we have the fast/slow setting but we then also offer a way to override the  those.  There is no place to override and set persistent buffering

alanconway replied:
fast implies in-memory. safe implies persistent buffers or end-to-end ack, whichever is most efficient given the source/sink combinations. Do we need more overrides?

kattz-kawa replied:
Please allow me to join this discussion abruptly, and let me share my thoughts.

One thing that comes to my mind as a concern is that, as indicated in "https://github.com/vectordotdev/vector/discussions/12991", 
end-to-end (e2e) support is not available in Vector for the combination of using Lua transfer (source -> lua transfer -> sink). Lua Transfer is used a part of OpenShift Logging.
Therefore it might be difficult to automatically determine e2e or Persistent buffer from only the combination of source and sink.

jcantrill replied:
@kattz-kawa we have a desire to refactor the lua transform and replace it with VRL.  I don't recall if we have a ticket to do that but I believe everything we need to do can now be accomplished with VRL.  If that is only blocker in your comment then we should proceed to prioritize the refactoring

kattz-kawa replied:
@jcantrill 
thank you for the reply. 
Remap transfer is supported for e2e, as far as I can see. 
so refactoring the lua transform is fine for my concern.

jcantrill replied:
@alanconway in summary of our f2f discussion, initial pass forgoes implicit disk buffering to use e2e acks which makes the original source "the buffer"

alanconway commented:
@jcantrill Great feedback, thanks. I'll have to do some re-writing to address it, see next push.

periklis commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,241 @@
+
+Should make existing failure modes more predictable and reliable.
+
+#### Support Procedures
+
+None.
```
As mentioned on the arch call, I suggest to add instructions of even better custom/recorded metrics on how to troubleshoot which output/sink is slowing down any other output/sinks at the source when using acks.

alanconway replied:
Noted in the enhancement, but no concrete metric design yet.

alanconway reviewed (commented):
@jcantrill @periklis please check if I've addressed your concerns.

periklis commented:
LGTM

cahartma commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,241 @@
+
+**Note**: The log collector _cannot_ guarantee reliable end-to-end delivery.
+It has no control over the reliability or throughput of sources and sinks.
+The `delivery` policy tunes the log collector's "best effort".
```
To Jeff's point below, I would add a note here pointing to the "implementation details", or maybe specifically mention the following list as "overrides".

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
````diff
@@ -0,0 +1,241 @@
+``` yaml
+outputs:
+  - name: detailed-tweaks
+    tuning:
````
This does not seem to reflect the rest of the document but adding comment here.

Maybe if the delivery implies memory or disk buffer then I still believe we need some way to further tune events and disk size. I would expect then we might provide something like:

```
tuning:
  delivery: fast|safe
  compression:
  request:
    maxDelay
  safe:
    maxEvents:
  fast:
    maxSize:    #this is bytes but could be '1G' which we convert ?
```

* I have comments down below regarding maxSize which I think we can calculate or make some assumptions and not allow it to be tweaked.
* I think we have to use vector term 'request' otherwise in future if we need to expose more, I don't see how the other parameters translate to 'reconnect'

cahartma commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,241 @@
+## Design Details
+
+### Open Questions
+None.
```
I anticipate a lot of questions about the details of the abstraction to fast/safe.   Would performance benchmarking be too obscure, or helpful showing examples of the implementation and its impact on a "test cluster"?    I'm assuming QE is going to need to verify the results somehow as well.

alanconway replied:
Check the updated text. The new text focuses on the reliability guarantees and only points out in passing that the unreliable delivery "does less bookkeeping and therefore uses less CPU/disk". I've de-emphasized the notion the trade-off is about performance, and the discussion is slated towards "You probably want reliability. If you are sure you don't then unreliable delivery *might* be a bit faster."

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
````diff
@@ -0,0 +1,241 @@
+``` yaml
+outputs:
+  - name: detailed-tweaks
+    tuning:
````
@cahartma transfering our discussion here:
 you mentioned in scrum that you were not fond of 'fast' and 'safe'.  Maybe we should consider an alternate to identify the expectation.  We are trying to make  an optimization between throughput and durability.  Maybe thats the proper identifier.  The intent I believe is to provide some "sane" defaults though without testing we really don't know what that is.  it could even differ based on output.  This means I think we need a third option IMO which is... tweak yourself.  so maybe 
```
tuning:
  limits:
     maxRecords: #int
  optimization:
    type: throughput|durability|manual #durable?
    batch:
      maxEvents:
      maxBytes:  #or some equivalent we could calc from something like 500Mb
    buffer:
      whenFull:  #block|dropNewest  or some equivalent
    request:
      maxDelay
    compression:   
```

alanconway replied:
Switched to standard messaging AtMostOnce/AtLeastOnce terminology. Not hugely intuitive outside the messaging world, but easy to explain and familiar to anyone who has reliable messaging experience.
I feel this is much better than fast/safe in many ways, thanks for pushing.

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Acknowledgements
+
+[End-to-end acknowledgement](https://vector.dev/docs/about/under-the-hood/architecture/end-to-end-acknowledgements) 
```
I think we need an explicit call out that we intend to implement via e2e ack instead of file buffering.  This describes e2e but does not really say that is how it is being implemented.  We otherwise risk the repeated question and inquiries that there is no buffering enabled.  Additionally, we need to test and verify that e2e acks=true is equivalent to e2e acks=false + buffering

alanconway replied:
Will have a think about how to phrase this, perhaps more explicit naming around acks vs. persistent buffers will solve the fast/safe naming problem.

kattz-kawa replied:
I would also suggest that rather than automatically enabling e2e, users should be able to choose whether to enable persistent buffering or let e2e do it for them.
This would also improve supportability in case e2e does not work.
When e2e does not work, the user can enable persistent buffering to achieve the purpose of delivery: slow.

alanconway replied:
@kattz-kawa agreed, my current thinking is to 
- enable e2e without buffering _by default_ - because it provides better reliability without much resource penalty
- allow manually forcing internal buffering as a user option, for special situations where e2e is inadequate
-  allow turning off e2e *without* buffering for special situations where reliability is not important and even the low cost of e2e is not wanted

alanconway replied:
See new draft: uses AtMostOnce, AtLeastOnce and AtLeastOncePersistent which I think is both clearer and  easier to explain. @jcantrill we can add AtLeastOncePersistent to your v2 API work or leave it for a later release.
It provides an escape hatch if e2e doesn't work in some situations, the implementation is just a combination of  e2e+persistent so probably not hard to add.

dhellmann commented:
#1555 is changing the enhancement template in a way that will cause the header check in the linter job to fail for existing PRs. If this PR is merged within the development period for 4.16 you may override the linter if the only failures are caused by issues with the headers (please make sure the markdown formatting is correct). If this PR is not merged before 4.16 development closes, please update the enhancement to conform to the new template.

alanconway commented:
Hopefully final draft: updated to be compatible with the v2 API proposal and @jcantrill PRs using AtLeastOnce/AtMostOnce instead of fast/safe. With the new terms, everything is easier to explain and I've re-directed the focus to reliability rather than performance which fixes some of @cahartma concerns. I've also added a "fallback" position to force use of persistence as suggested by @kattz-kawa.

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
````diff
@@ -0,0 +1,260 @@
+``` yaml
+outputs:
+  - name: detailed-tweaks
+    tuning:
````
Needs update to match https://github.com/openshift/cluster-logging-operator/blob/master/api/logging/v1/cluster_log_forwarder_types.go#L234-L269

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,260 @@
+Acknowledgements give similar reliability to persistent buffering but are more efficient.
+A persistent buffer is used if acknowledgements are not available.
+
+##### `AtLeastOncePersistent`
```
Is this something we "may" implement in the future?  I would strongly urge we consider dropping this until we are certain we actually need buffering.  Actual buffering has significant implications for the implementation given we currently (in fluentd) carve out some of the host for our own use.  This may actually require we have some persistent storage.  Consider hiding the text in the rendered output

alanconway replied:
+1 but Note that we already need buffering for cases where acks are not available, which sadly seems to include pod logs :disappointed:  so it's not adding much. I'm OK with dropping this option for the first cut if you think it's not necessary.

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,260 @@
+
+##### `AtLeastOncePersistent`
+
+Like `AtLeastOnce`, but forces the use of a persistent buffer, possibly in addition to acknowledgements.
```
Do we want to change the word "buffer" for "internal event queue" or similar?  "Buffer" has the connotation of a permanent store  where logs will never be lost.

alanconway replied:
I do mean a persistent buffer here - the idea is this option _forces_ the use of a persistent buffer, even if acks are available. The use case would be that the source of logs theoretically support acks, but the source itself is not reliable, so even though it can ack, it's not trustworthy to replay data. We can leave this option out of first cut if you think that's best.

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,260 @@
+- Value indicates compression type (e.g. "zlib", "gzip"), valid values depend on the output type.
+- Error if output does not support compression or does not recognize the value.
+
+#### batch: object
```
Does match the impl API

alanconway replied:
Fixed

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,260 @@
+- `maxRecords`: Max number of records to include in a batch.
+- `maxBytes`: Max size in bytes of a batch
+
+#### reconnect: object
```
Does match the impl API

jcantrill commented on enhancements/cluster-logging/performance-tuning.md:
```diff
@@ -0,0 +1,260 @@
+means deliberately dropping log records from the collector.
+Review the existing limit code when implementing delivery policy so that the two work together properly.
+
+### Risks and Mitigations
```
AtLeastOncePersistent has significant risk related to the actual storage to be implemented when considering its usage in mCLF

alanconway replied:
+1, see previous comment

jcantrill reviewed (commented):
/hold

alanconway commented:
/approve
